Makefile

```diff
@@ -66,91 +55,76 @@ endif
 	hack/verify-generated-swagger-spec.sh
 .PHONY: verify
 
-# check and verify can't run concurently because of strange concurrent build issues.
-check-verify:
-	# delegate to another make process that runs serially against the check and verify targets
-	$(MAKE) -j1 check verify
-.PHONY: check-verify
-
-# Install travis dependencies
-#
-# Args:
-#   TEST_ASSETS: Instead of running tests, test assets only.
-ifeq ($(TEST_ASSETS), true)
-install-travis:
-	hack/install-assets.sh
-else
-install-travis:
-	hack/install-etcd.sh
-	hack/install-tools.sh
-endif
-.PHONY: install-travis
-
-# Run unit and integration tests that don't require Docker.
+# Run unit tests.
 #
 # Args:
+#   WHAT: Directory names to test.  All *_test.go files under these
+#     directories will be run.  If not specified, "everything" will be tested.
+#   TESTS: Same as WHAT.
 #   GOFLAGS: Extra flags to pass to 'go' when building.
 #   TESTFLAGS: Extra flags that should only be passed to hack/test-go.sh
-#   TEST_ASSETS: Instead of running tests, test assets only.
 #
 # Example:
-#   make check-test
-check-test: export KUBE_COVER= -cover -covermode=atomic
-check-test: export KUBE_RACE=  -race
-ifeq ($(TEST_ASSETS), true)
-check-test:
-	hack/test-assets.sh
-else
-check-test: 
-	check-verify
-	hack/test-cmd.sh
-	KUBE_RACE=" " hack/test-integration.sh
-endif
-.PHONY: check-test
+#   make test-unit
+#   make test-unit WHAT=pkg/build GOFLAGS=-v
+test-unit:
+	TEST_KUBE=1 hack/test-go.sh $(WHAT) $(TESTS) $(TESTFLAGS)
+.PHONY: test-unit
 
-# Build and run the complete test-suite.
-#
-# Args:
-#   GOFLAGS: Extra flags to pass to 'go' when building.
-#   TESTFLAGS: Extra flags that should only be passed to hack/test-go.sh
+# Run integration tests. Compiles its own tests, cannot be run
+# in parallel with any other go compilation.
 #
 # Example:
-#   make test
-#   make test GOFLAGS=-v
-test: export KUBE_COVER= -cover -covermode=atomic
-test: export KUBE_RACE=  -race
-ifeq ($(SKIP_BUILD), true)
-test: check-verify test-int-plus
-else
-test: build check-verify test-int-plus
-endif
-.PHONY: test
+#   make test-integration
+test-integration:
+	KUBE_COVER=" " KUBE_RACE=" " hack/test-integration.sh
+.PHONY: test-integration
 
-# Split out of `test`.  This allows `make -j --output-sync=recurse test` to parallelize as expected
-test-int-plus: export KUBE_COVER= -cover -covermode=atomic
-test-int-plus: export KUBE_RACE=  -race
-ifeq ($(SKIP_BUILD), true)
-test-int-plus: 
-else
-test-int-plus: build
-endif
-test-int-plus:
+# Run command tests. Uses whatever binaries are currently built.
+#
+# Example:
+#   make test-cmd
+test-cmd: build
 	hack/test-cmd.sh
+.PHONY: test-cmd
+
+# Run end to end tests. Uses whatever binaries are currently built.
+#
+# Example:
+#   make test-cmd
+test-end-to-end: build
+	hack/test-end-to-end.sh
+.PHONY: test-end-to-end
+
+# Run tools tests.
```

Comment:
I'm about to add `hack/test-lib.sh` to test our bash libraries. Does that deserve it's own make target? How do we make that choice?